rfcs/v3/authorization-rules.md

```diff
@@ -0,0 +1,568 @@
+
+For example, if I have models for `Channels`, `Threads`, `Replies`. Then my predicates would be as follows:
+- `Channels`: `members._contains(x-hasura-user-id)`
+- `Threads`: `channel.members._contains(x-hasura-user-id`
+- `Replies`: `thread.channel.members._contains(x-hasura-user-id)`.
+
+Ideally, you would be able to express the fact that an object should be selectable if the related object is also selectable.
+
+Hence, we propose the following addition to the permissions predicate:
+```yaml
+allowObjects:
+  relationship:
+    name: 'my_relationship'
+    relatedObjectAllowed: true
+```
+
+### AuthConfig
+
+The notion of roles will be removed from AuthConfig. In particular, the `allow_role_emulation_by` key will be replaced by a `allow_session_emulation_if` key which specifies a condition which needs to evaluate to true for the request to allow emulating session variables.
+
+No `x-hasura-` session variables will now be required in the webhook response or in JWT claims. Users can choose arbitrary names for session variables.
+
+### Implementation
+
+The above proposal has major implications for engine, the `lang-graphql` crate in particular. Currently, `lang-graphql` allows constructing a single graphql schema with information about all roles embedded into the schema and allows query validation / IR generation to happen in the context of a particular role only. However, with the above proposal, there is no notion of roles anymore. So, `lang-graphql` will need to move to an architecture where when trying to access a graphql field, instead of looking up an allowed set of roles, it can evaluate an arbitrary expression that evaluates to a boolean. The details of this architecture are out of scope for this RFC.
+
+#### Milestones
+
+- Milestone 1: Refactor the internal engine implementation to support the new permissions system without any metadata changes. Internally, the role-based definitions would be mapped to the new system.
+- Milestone 2: Make the metadata changes for the new system with the following primitives:
+  - Conditional activation of rules. Conditions support a basic set of operators (`equals`, `isNull`, and logical operators at least).
+  - Value expressions can either be literals or session variables only.
+  - Support only field permissions / model select predicates / argument presets. No input validation.
+  - Composability of rules.
+- New features (in no particular order):
+  - Reusing relationship permissions
+  - Arguments input validation
+  - Calling external systems
+  - Accessing sub-fields in value expressions
+  - Advanced boolean operators
+
+## FAQs
+
+1. **Will existing metadata continue to be work?**  
+    
+    Yes, existing metadata (using the `v1` version of permissions objects) will continue to work and trivially maps to the new system where the condition is an equality check on the `x-hasura-role` session variable.
```

Comment:
I think we should consider letting the user pick which way they'd prefer to express their permissions, either this advanced system, or the existing role based system (which we just desugar into the the advanced system behind the scenes). This advanced system is very powerful, but also less obvious how to use it to recreate the simple use case and I contend that users who just need basic role-based permissions would find the old mechanism easier to author and understand. 

To be clear, this doesn't mean just retaining support for the v1 ModelPermissions and TypePermissions, it means including both options on v2. Just retaining the old v1 types is not as good as it implicitly discourages their use because they are an "old version".